File: go/pricing/pricesummary_test.go
// ABOUTME: Unit tests for the PriceSummary domain type.
// ABOUTME: Verifies price summary creation and field assignment.
package pricing

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"promotionengine/domain"
)

func TestNewPriceSummary(t *testing.T) {
	subtotal := domain.Euros(10.00)
	discountTotal := domain.Euros(2.00)
	total := domain.Euros(8.00)

	discount := NewAppliedDiscount(
		domain.MustPromotionID("PROMO10"),
		domain.Euros(2.00),
		"cart",
		"10% off",
	)
	appliedDiscounts := []AppliedDiscount{discount}

	summary := NewPriceSummary(subtotal, discountTotal, total, appliedDiscounts)

	assert.Equal(t, subtotal, summary.Subtotal)
	assert.Equal(t, discountTotal, summary.DiscountTotal)
	assert.Equal(t, total, summary.Total)
	assert.Len(t, summary.AppliedDiscounts, 1)
	assert.Equal(t, discount, summary.AppliedDiscounts[0])
}

func TestNewPriceSummaryNoDiscounts(t *testing.T) {
	subtotal := domain.Euros(10.00)
	discountTotal := domain.Euros(0.00)
	total := domain.Euros(10.00)
	appliedDiscounts := []AppliedDiscount{}

	summary := NewPriceSummary(subtotal, discountTotal, total, appliedDiscounts)

	assert.Equal(t, subtotal, summary.Subtotal)
	assert.Equal(t, discountTotal, summary.DiscountTotal)
	assert.Equal(t, total, summary.Total)
	assert.Empty(t, summary.AppliedDiscounts)
}

File: typescript/src/pricing/PriceSummary.test.ts
import { describe, it, expect } from 'vitest';
import { PriceSummary } from './PriceSummary';
import { AppliedDiscount } from './AppliedDiscount';
import { Money } from '../domain/Money';
import { PromotionId } from '../domain/PromotionId';

describe('PriceSummary', () => {
  describe('construction', () => {
    it('should construct with no applied discounts', () => {
      const subtotal = Money.euros('45.00');
      const discountTotal = Money.euros('0.00');
      const total = Money.euros('45.00');
      const summary = new PriceSummary(subtotal, discountTotal, total, []);

      expect(summary.subtotal).toBe(subtotal);
      expect(summary.discountTotal).toBe(discountTotal);
      expect(summary.total).toBe(total);
      expect(summary.appliedDiscounts).toHaveLength(0);
    });

    it('should construct with applied discounts', () => {
      const subtotal = Money.euros('45.00');
      const discountTotal = Money.euros('5.00');
      const total = Money.euros('40.00');
      const discount = new AppliedDiscount(
        new PromotionId('PROMO-123'),
        Money.euros('5.00'),
        'cart',
        'Test discount'
      );
      const summary = new PriceSummary(subtotal, discountTotal, total, [discount]);

      expect(summary.subtotal).toBe(subtotal);
      expect(summary.discountTotal).toBe(discountTotal);
      expect(summary.total).toBe(total);
      expect(summary.appliedDiscounts).toHaveLength(1);
      expect(summary.appliedDiscounts[0]).toBe(discount);
    });
  });

  describe('immutability', () => {
    it('should not allow external modification of applied discounts', () => {
      const discount = new AppliedDiscount(
        new PromotionId('PROMO-123'),
        Money.euros('5.00'),
        'cart',
        'Test discount'
      );
      const summary = new PriceSummary(
        Money.euros('45.00'),
        Money.euros('5.00'),
        Money.euros('40.00'),
        [discount]
      );

      // TypeScript should prevent this at compile time,
      // but we test runtime immutability
      expect(() => {
        // @ts-expect-error - Testing runtime immutability
        summary.appliedDiscounts.push(discount);
      }).toThrow();
    });
  });
});
